### DISCUSSION TITLE
Is there any way to return dynamic pages?

### DISCUSSION THREAD

**Discussion by whispersilk** [tauri-apps/tauri#11387] [Q&A]
I know that it's possible to build multi-page applications in Tauri, but I'm having a hard time figuring out if there is a way to render a dynamic page in Rust before returning it, as you might in a traditional web application. Is this possible? To be clear, what I mean is that instead of doing something like:

```html
<!-- /path/page.html -->

<!doctype HTML>
<html>
    <head></head>
    <body>
        <script type="text/javascript">
            document.innerHTML = window.__TAURI__.core.invoke("dynamic_render", { path: "/path/page" });
        </script>
    </body>
</html>
```

is it possible to do something like:

```rust
// page.rs

fn register_handler(tauri: &mut tauri::Builder) {
    tauri.register("/path/page.html", render_page);
}

async fn render_page() -> String {
    String::from(r#"
        <some html content>
    "#)
}
```

Would [register_asynchronous_uri_scheme_protocol](https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_asynchronous_uri_scheme_protocol) be a good way to accomplish this (with pages linking to `dummy-protocol:///path/page.html` instead of `/path/page.html`)? Or is there a better/other way?

**Comment by FabianLars** [comment]
> Would [register_asynchronous_uri_scheme_protocol](https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.register_asynchronous_uri_scheme_protocol) be a good way to accomplish this (with pages linking to dummy-protocol:///path/page.html instead of /path/page.html)?

Yep, that's pretty good imo. Alternatively you could use a http server but that would of course expose your frontend on the whole device network if that's a concern.
